### ISSUE TITLE
C/C++ Intellisense crashes on my template code.

### ISSUE THREAD

**Issue by rlj1202** [microsoft/vscode-cpptools#7349]
Bug type: Language Service

**Describe the bug**
- OS and Version: Linux codespaces_56f35f 5.4.0-1043-azure #45~18.04.1-Ubuntu SMP Sat Mar 20 16:16:05 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
- VS Code Version: 1.55.1
- C/C++ Extension Version: 1.2.2
- Other extensions you installed (and if the issue persists after disabling them): CMake, CMake Tools, Clang-Format, crates, Git Graph, GitHub Theme, Live Share, Rust, rust-analyzer, Visual Studio IntelliCode, XML Tools
- If using SSH remote, specify OS of remote machine: Uhm... same as above? Am using Github Codespace on browser.
- A clear and concise description of what the bug is, including information about the workspace (i.e. is the workspace a single project or multiple projects, size of the project, etc).

**Steps to reproduce**

Create cpp file and copy-paste following code occurs the error.

**Expected behavior**

After I removed following code, intellisense works perfectly.

**Code sample and logs**

- Code sample
```c++
template<typename Datatype, int... Ptrs>
class Vbo {
};

template <typename... Vbos>
class _Mesh {
private:
    template<typename DataType>
    class GLType {
    public:
        static const unsigned int value;
    };

    template<>
    class GLType<float> { public: static const unsigned int value = GL_FLOAT; };
    template<>
    class GLType<int> { public: static const unsigned int value = GL_INT; };

    template<typename... Ts>
    class VboIter {
    public:
        void CreateVbos(int loc);
        void DeleteVbos();
    };

    template<>
    class VboIter<> {
    public:
        void CreateVbos(int loc) {}
        void DeleteVbos() {}
    };

    template<typename DataType, int... Ptrs, typename... Rests>
    class VboIter<Vbo<DataType, Ptrs...>, Rests...> : public VboIter<Rests...> {
    public:
        template<typename Type, typename... Types>
        void CreateVbos(int loc, Type data, Types... rests) {
            glGenBuffers(1, &vbo_id);
            glBindBuffer(GL_ARRAY_BUFFER, vbo_id);
            glBufferData(GL_ARRAY_BUFFER, sizeof(DataType) * data.size(), (void*) &data[0], GL_STATIC_DRAW);

            IteratePtrs(loc, 0, Ptrs...);

            VboIter<Rests...>::CreateVbos(loc + sizeof...(Ptrs), rests...);
        }

        void DeleteVbos() {
            glDeleteBuffers(1, &vbo_id);

            VboIter<Rests...>::DeleteVbos();
        }

        void CreatePtr(int loc, int offset, int size) {
            unsigned int type = GLType<DataType>::value;
            int total = (Ptrs + ...);

            glVertexAttribPointer(
                loc, size, type, GL_FALSE,
                sizeof(DataType) * total, (void*) (sizeof(DataType) * offset)
            );
            glEnableVertexAttribArray(loc);

            DEBUG_STDOUT("Set vertex pointer : loc = %d, offset = %d, size = %d\n", loc, offset, size);
        }

        void IteratePtrs(int loc, int offset, int size) {
            CreatePtr(loc, offset, size);
        }

        void IteratePtrs(int loc, int offset, int size, int rests...) {
            CreatePtr(loc, offset, size);

            IteratePtrs(loc + 1, offset + size, rests);
        }

        unsigned int vbo_id;
    };

public:
    template<typename... Types>
    _Mesh(std::vector<unsigned int> indices, Types... data) {
        glGenVertexArrays(1, &vao_id);
        glBindVertexArray(vao_id);

        vbo_iter.CreateVbos(0, data...);

        glGenBuffers(1, &ebo_id);
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, ebo_id);
        glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(unsigned int) * indices.size(), (void*) &indices[0], GL_STATIC_DRAW);

        cnt_vertices = indices.size();

        glBindVertexArray(0);
    }
    ~_Mesh() {
        if (vao_id) {
            glDeleteVertexArrays(1, &vao_id);
        }
        vbo_iter.DeleteVbos();
        if (ebo_id) {
            glDeleteBuffers(1, &ebo_id);
        }
    }

    void Bind() {
        glBindVertexArray(vao_id);
    }

    void Draw() {
        glDrawElements(GL_TRIANGLES, cnt_vertices, GL_UNSIGNED_INT, (void*) 0);
    }

private:
    unsigned int vao_id;
    unsigned int ebo_id;
    unsigned int cnt_vertices;

    VboIter<Vbos...> vbo_iter;
};
```
- Configurations in `c_cpp_properties.json`
```json
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [
                "EMSCRIPTEN"
            ],
            "cStandard": "c17",
            "cppStandard": "c++17",
            "compileCommands": "${workspaceFolder}/build/compile_commands.json",
            "configurationProvider": "ms-vscode.cmake-tools",
            "browse": {
                "path": [
                    "/usr/include",
                    "/workspaces/codespace/emscripten/system/include"
                ]
            }
        }
    ],
    "version": 4
}
```

- Logs from [the language server logging](https://code.visualstudio.com/docs/cpp/enable-logging-cpp#_enable-logging-for-the-language-server)
```
완료 제공 중
서버에 1 변경 내용을 보내는 중
completionItem/resolve (id: 1301)
cpptools/getFoldingRanges: file:///workspaces/codespace/template_test/main.cpp (id: 1302)
cpptools/getCodeActions: file:///workspaces/codespace/template_test/main.cpp (id: 1303)
textDocument/didChange: file:///workspaces/codespace/template_test/main.cpp
cpptools/textEditorSelectionChange
cpptools/getFoldingRanges: file:///workspaces/codespace/template_test/main.cpp (id: 1304)
cpptools/fileChanged: file:///workspaces/codespace/template_test/main.cpp
textDocument/didSave: file:///workspaces/codespace/template_test/main.cpp
  태그 구문 분석 파일: /workspaces/codespace/template_test/main.cpp
cpptools/getDocumentSymbols: file:///workspaces/codespace/template_test/main.cpp (id: 1307)
cpptools/getDocumentSymbols
cpptools/fileChanged: file:///workspaces/codespace/template_test/main.cpp
textDocument/didSave: file:///workspaces/codespace/template_test/main.cpp
구문 오류를 확인하는 중: file:///workspaces/codespace/template_test/main.cpp
IntelliSense 프로세스 크래시가 감지되었습니다. <- IntelliSense process crash is detected
IntelliSense 시간(초) 업데이트: 3.259
유휴 루프: 활성 문서를 구문 분석하는 중
구문 오류를 확인하는 중: file:///workspaces/codespace/template_test/main.cpp
IntelliSense 시간(초) 업데이트: 0
```

**Screenshots**

![screenshot](https://user-images.githubusercontent.com/10295515/114489385-03cbc480-9c4e-11eb-9408-bd420bce122a.png)

**Additional context**

Regardless the GL header file is not include, the problem occurs in a same manner. I remember that the code was intellisensed before correctly but at some point, this thing have been happened.

If there is anything I need to provide, I am willing to do so.

**Comment by sean-mcmanus** [comment]
Thanks for reporting this. I've filed a bug on our shared VS parser at https://developercommunity.visualstudio.com/t/C-IntelliSense-processes-crashing-when/1397318

**Comment by sean-mcmanus** [comment]
Looks like it's being tracked by https://developercommunity.visualstudio.com/t/Broken-intellisense-with-template-specia/1359157 now.

**Comment by Colengms** [comment]
This should be addressed in 1.6.0-insiders.
